sql/src/main/java/io/crate/expression/reference/doc/lucene/DocCollectorExpression.java

```diff
@@ -99,7 +98,34 @@ public Object value() {
             // for example:
             //      sourceExtractor might read byte as int and
             //      then eq(byte, byte) would get eq(byte, int) and fail
-            return returnType.value(sourceLookup.extractValue(columnName));
+            return returnType.value(extractValue(sourceLookup, 0, path));
+        }
+
+        @SuppressWarnings("unchecked")
+        private static Object extractValue(final Map map, int index, List<String> path) {
+            assert path instanceof RandomAccess : "path should support RandomAccess for fast index optimized loop";
+            Map m = map;
+            Object tmp = null;
+            for (int i = index; i < path.size(); i++) {
+                tmp = m.get(path.get(i));
+                if (tmp instanceof Map) {
+                    m = (Map) tmp;
+                } else if (tmp instanceof List) {
+                    List list = (List) tmp;
+                    List newList = new ArrayList(list.size());
+                    for (Object o : list) {
+                        if (o instanceof Map && i + 1 < path.size()) {
+                            newList.add(extractValue((Map) o, i + 1, path));
```

Comment:
since this will be called for every row, should we drop the recursion and replace it with a while loop to avoid the stack allocations?